### DISCUSSION TITLE
Touch Sensors not always registering contact

### DISCUSSION THREAD

**Discussion by kuds** [google-deepmind/mujoco#2483] [Asking for Help] [answered]
### Intro

Hi!

I am an independent researcher interested in using reinforcement learning to solve various problems and to build experiments in MujoCo.

### My setup

Google Colab T4
Python Version: 3.11.11
Torch Version: 2.5.1+cu124
Is Cuda Available: True
Cuda Version: 12.4
Gymnasium Version: 1.0.0
Numpy Version: 1.26.4
Mujoco Version: 3.3.0
Stable-Baselines3 Version: 2.5.0
Matplotlib Version: 3.10.0

### My question

I am trying to build a reinforcement learning system that learns to bounce a ball using Gymnasium, Stable Baselines3, and MuJoCo. In reviewing the simulation recording and contact data, there appears to be a misalignment between what the video shows and how many times the touch sensor registers a force. The embedded video shows at least six (6) bounces, but the sensor data only lists three (3) (see attached CSV).

[ball_bounce_200000.csv](https://github.com/user-attachments/files/19134127/ball_bounce_200000.csv)

![ball_bounce_200000-step-0-to-step-1000018-ezgif com-video-to-gif-converter](https://github.com/user-attachments/assets/0c1fa341-5806-4c2d-a58f-d11f9aa629d6)

**Relevant Issues:**
* https://github.com/google-deepmind/mujoco/issues/228
* https://github.com/google-deepmind/mujoco/issues/1823

The above issues stem from the touch sensor area being too small. So, I increased the site size to include the entire paddle, but the touch sensor does not always appear to register contact.

**Relevant Documentation:**
* https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-touch
* https://mujoco.readthedocs.io/en/latest/computation/index.html

What is the reason why MuJoCo is not registering all contact instances via the touch sensor? Is it because the timestep is too big, or how the solref is defined for the paddle and ball geom? 

Any help would be greatly appreciated!

## Link to Google Colab Notebook
[[Ball Bounce] Soft Actor-Critic (SAC)](https://github.com/kuds/rl-mujoco-tennis/blob/main/%5BBall%20Bounce%5D%20Soft%20Actor-Critic%20(SAC).ipynb)

## MuJoCo World Model
```xml
<mujoco>
  <option timestep=".002" />
  <asset>
    <texture name="grid" type="2d" builtin="checker" rgb1=".1 .2 .3"
      rgb2=".2 .3 .4" width="300" height="300" mark="none" />
    <material name="grid" texture="grid" texrepeat="1 1"
      texuniform="true" reflectance=".2" />
  </asset>
  <worldbody>
    <light name="light" pos="0 0 1" />
    <geom name="floor" type="plane" pos="0 0 -0.5" size="2 2 .1" material="grid" />
    <body name="paddle_body" pos="0 0 0">
      <site name="bounce_sensor" pos="0 0 0.025" size="0.21 0.21 0.075" type="box" rgba="1 0 0 .05" />
      <joint name="rotate_x" pos="0 0 0" axis="1 0 0" range="-.3 .3" damping="10" />
      <joint name="rotate_y" pos="0 0 0" axis="0 1 0" range="-.3 .3" damping="10" />
      <joint name="rotate_z" pos="0 0 0" axis="0 0 1" range="-.3 .3" damping="10" />
      <joint axis="0 0 1" limited="true" name="slider_z" range="0 0.5" type="slide" />
      <joint axis="0 1 0" limited="true" name="slider_y" range="-.2 .2" type="slide" />
      <joint axis="1 0 0" limited="true" name="slider_x" range="-.2 .2" type="slide" />
      <geom name="paddle_geom" type="box" size=".2 .2 .02" rgba="0.2 0.2 1 1" solref="0.02 0.07" />
    </body>
    <body name="ball_body" pos="0 0 0.3">
      <freejoint name="ball_freejoint" />
      <geom name="ball_geom" type="sphere" size=".04" rgba="0 1 0 1" solref="0.02 0.07" />
    </body>
  </worldbody>
  <actuator>
    <general name="act_x" joint="rotate_x" ctrlrange="-1 1" gainprm="500 0 0" />
    <general name="act_y" joint="rotate_y" ctrlrange="-1 1" gainprm="500 0 0" />
    <general name="act_z" joint="rotate_z" ctrlrange="-1 1" gainprm="500 0 0" />
    <motor ctrllimited="true" ctrlrange="-1 1" gear="50" joint="slider_z" name="slider_z_motor" />
    <motor ctrllimited="true" ctrlrange="-1 1" gear="50" joint="slider_y" name="slider_y_motor" />
    <motor ctrllimited="true" ctrlrange="-1 1" gear="50" joint="slider_x" name="slider_x_motor" />
  </actuator>
  <sensor>
    <touch name="touch_sensor" site="bounce_sensor" />
  </sensor>
</mujoco>
```

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment]
It's a bit hard to see in your video, but I think the touch sensor is not properly aligned, it seems to be above the box rather than enclosing it. If you have a deep penetration then the position of the contact could easily be quite deep in the paddle, which would not be caught by the sensor.

**Comment by yuvaltassa** [comment]
PS, you might want to consider using a fromto sensor, and registering a collision when the fromto vector flips direction

**Comment by kuds** [comment]
@yuvaltassa 

Thanks for the advice! I added a `fromto`, `accelerator`, and `velocimeter` sensors to the environment and made the touch sensor site more visible in the simulations. 

## Updated Results

###  300,000 Training Steps 

[ball_bounce_300000.csv](https://github.com/user-attachments/files/19172688/ball_bounce_300000.csv)

![image](https://github.com/user-attachments/assets/8fe2b05e-d9d7-4a43-95ec-6d2e232e52ff)

https://github.com/user-attachments/assets/75c1b8ba-a9cd-40af-b468-b2883a113238

### Best Model (500,000 Training Steps)

[best_model_ball_bounce.csv](https://github.com/user-attachments/files/19172687/best_model_ball_bounce.csv)

![image](https://github.com/user-attachments/assets/65e5477c-4473-4934-9952-b9378de6f0cd)

https://github.com/user-attachments/assets/e02bbbc3-265d-417c-a638-df3ffa1cbbc8

### Updated MuJoCo Model
```xml
<mujoco>
  <option timestep=".002" />
  <asset>
    <texture name="grid" type="2d" builtin="checker" rgb1=".1 .2 .3"
      rgb2=".2 .3 .4" width="300" height="300" mark="none" />
    <material name="grid" texture="grid" texrepeat="1 1"
      texuniform="true" reflectance=".2" />
  </asset>
  <worldbody>
    <light name="light" pos="0 0 1" />
    <geom name="floor" type="plane" pos="0 0 -0.5" size="2 2 .1" material="grid" />
    <body name="paddle_body" pos="0 0 0">
      <site name="bounce_sensor" pos="0 0 0.025" size="0.21 0.21 0.075" type="box" rgba="1 0 0 .15" />
      <joint name="rotate_x" pos="0 0 0" axis="1 0 0" range="-.3 .3" damping="10" />
      <joint name="rotate_y" pos="0 0 0" axis="0 1 0" range="-.3 .3" damping="10" />
      <joint name="rotate_z" pos="0 0 0" axis="0 0 1" range="-.3 .3" damping="10" />
      <joint axis="0 0 1" limited="true" name="slider_z" range="0 0.5" type="slide" />
      <joint axis="0 1 0" limited="true" name="slider_y" range="-.2 .2" type="slide" />
      <joint axis="1 0 0" limited="true" name="slider_x" range="-.2 .2" type="slide" />
      <geom name="paddle_geom" type="box" size=".2 .2 .02" rgba="0.2 0.2 1 1" solref="0.02 0.07" />
    </body>
    <body name="ball_body" pos="0 0 0.3">
      <freejoint name="ball_freejoint" />
      <site name="ball_site" pos="0 0 0" size="0.04" rgba="0 1 0 1" />
      <geom name="ball_geom" type="sphere" size=".04" rgba="0 1 0 1" solref="0.02 0.07" />
    </body>
  </worldbody>
  <actuator>
    <general name="act_x" joint="rotate_x" ctrlrange="-1 1" gainprm="500 0 0" />
    <general name="act_y" joint="rotate_y" ctrlrange="-1 1" gainprm="500 0 0" />
    <general name="act_z" joint="rotate_z" ctrlrange="-1 1" gainprm="500 0 0" />
    <motor ctrllimited="true" ctrlrange="-1 1" gear="50" joint="slider_z" name="slider_z_motor" />
    <motor ctrllimited="true" ctrlrange="-1 1" gear="50" joint="slider_y" name="slider_y_motor" />
    <motor ctrllimited="true" ctrlrange="-1 1" gear="50" joint="slider_x" name="slider_x_motor" />
  </actuator>
  <sensor>
    <fromto name="ball_to_paddle" body1="ball_body" body2="paddle_body" />
    <accelerometer site="ball_site" name="ball_accelerometer" />
    <velocimeter site="ball_site" name="ball_velocity" />
    <touch name="touch_sensor" site="bounce_sensor" />
  </sensor>
</mujoco>
```

The ball's velocity is aligned with the recorded simulation. However, the `fromto` or the `accelerator` sensors only appear to measure significantly high forces and do not capture small interactions. Is this to be expected, or is something wrong with my MuJoCo XML configuration? 

I appreciate your help with this! Let me know if you need any other details or want to see any specific code!

**Reply by yuvaltassa** [reply, marked as answer]
I don't think there is a problem, everything is measured. The contact forces and accelerations are simply very high (for a short span of time), so the smaller forces/accelerations related to the smooth motions aren't visible in your auto-scaled plots. If you zoom in the y-axis you will see all the details...

**Reply by kuds** [reply]
Understood! I am still learning the intricacies of MuJoCo and how different sensors log their respective measurements. 

Thanks again for the help!
